### ISSUE TITLE
Handling on a server connections closed by clients

### ISSUE THREAD

**Issue by ruippeixotog** [akka/akka-http#1718]
I have an HTTP service written with akka-http that, due to legacy API compatibility, has some synchronous endpoints for which requests can take a long time to be processed. The server is supposed to keep the connection open during that time.

Due to the long time those requests can take, clients sometimes close connections prematurely (most probably due to HTTP client timeouts). When that happens, I'd like to handle the closed connection by cancelling the processing for that request - otherwise, the server will get stuck with a backlog of processing tasks that won't be used later. I can't seem to find a way to do that.

**Comment by jrudolph** [comment]
Yes, indeed that's currently hard to respond to. You can get the information that the client has closed the connection by providing a `Flow[HttpRequest, HttpResponse, _]` and watching for termination / failure of the source side. The standard implementation of this flow in the routing layer, however, does not watch for that information.

What we could do is change `Route.handlerFlow` to add a `watchTermination` stage and through that provide a `Future[Done]` to the `RequestContext` that could be inquired about the current state of the connection.

**Comment by ruippeixotog** [comment]
That's right, I was able to do that now by wrapping the `Flow` of requests (I tried that before, but I couldn't get the `watchTermination` part right). For anyone interested, here's a snippet of the server binding code:

```scala
val serverSource: Source[Http.IncomingConnection, Future[Http.ServerBinding]] =
  Http(context.system).bind(interface = "0.0.0.0", port = 8080)

serverSource.to(Sink.foreach { connection =>
  println("Accepted new connection from " + connection.remoteAddress)

  // materializes to a Future[Done] completed when the connection is closed
  connection.flow.watchTermination()(Keep.right)
    .joinMat(serviceRoutes)(Keep.left).run()
}).run()
```

Having that `Future[Done]` accessible through the `RequestContext` would be great, as it would be seamlessly passed through the route DSL. If you agree with having this, I can submit a PR with it.

Meanwhile, is there any place inside the context where we can pass things like this, and possibly other metadata bound to the request?

**Comment by jrudolph** [comment]
> If you agree with having this, I can submit a PR with it.

Yep, looks good.

> Meanwhile, is there any place inside the context where we can pass things like this, and possibly other metadata bound to the request?

For now we could add another field to the `RequestContext`. We want to introduce typed request/response attributes for cases like this to add internal metadata.

**Comment by KenCoder** [comment]
I'm having the same problem - @ruippeixotog did you ever submit that PR? I'm switching from bindAndHandle to bind so I can recognize disconnections, but I'm having to give up some of what bindAndHandle does.

**Comment by ruippeixotog** [comment]
@KenCoder sorry for the long delay in the response. I tried to do it at the time, but I found out that the current APIs didn't allow me to do that as easily as I could. Unfortunately, I cannot remember the details... If I remember correctly, it had to do with a dependency cycle in the creation of the instances, but you should definitely check this up again :)
